add support for the l4-bender linker on the x86_64-unknown-l4re-uclibc tier 3 target

atopia opened the pull request:
This PR contains the work by @humenda to update support for the `x86_64-unknown-l4re-uclibc` tier 3 target (published at [humenda/rust](https://github.com/humenda/rust)), rebased and adapted to current rust in follow up commits by myself. The publishing of the rebased changes is authorized and preferred by the original author. As the goal was to distort the original work as little as possible, individual commits introduce changes that are incompatible to the newer code base that the changes were rebased on. These incompatibilities have been remedied in follow up commits, so that the PR as a whole should result in a clean update of the target.
If you prefer another strategy to mainline these changes while preserving attribution, please let me know.

rust-highfive commented:
Some changes occured to rustc_codegen_cranelift

cc @bjorn3

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @matthewjasper (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

bjorn3 commented on library/std/src/sys/unix/l4re.rs:
```diff
@@ -481,7 +481,7 @@ pub mod net {
     impl LookupHost {
         pub fn port(&self) -> u16 {
-            unimpl!();
+            0 // unimplemented
```
Does this code actually get reached in your programs? Keeping `unimpl!()` until it is actually implemented will make it very clear when you accidentally depend on it, instead of potentially getting confused by a 0 value being returned for some reason.

atopia replied:
No, this code only exists to provide a uniform interface on "unix" platforms. The problem is that `port()` returns a `u16`, so the `Err()` returned by `unimpl!()` can't be used here. While the mock implementation could follow the structure of other platforms without networking support more closely, returning `0` is more or less in line with what other platforms such as [sgx](https://github.com/rust-lang/rust/blob/595088d602049d821bf9a217f2d79aea40715208/library/std/src/sys/sgx/net.rs#L473) or [wasi](https://github.com/rust-lang/rust/blob/595088d602049d821bf9a217f2d79aea40715208/library/std/src/sys/wasi/net.rs#L350) do. Since the commit only fixes up the most pressing issues with the existing mainline implementation, my suggestion would be to leave the `0` for now.

bors commented:
:umbrella: The latest upstream changes (presumably #79608) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

petrochenkov commented on compiler/rustc_codegen_ssa/src/back/linker.rs:
```diff
@@ -1284,6 +1286,180 @@ impl<'a> Linker for WasmLd<'a> {
+    fn control_flow_guard(&mut self) {}
+
+    fn no_crt_objects(&mut self) {}
+}
```
The interface looks like more or less like an `ld` linker, I'd suggest to avoid adding a new linker flavor for a start and reuse `GccLinker` (especially given that the set of linker flavors accepted by `-C linker-flavor` is a stable public interface).

humenda commented:
>The interface looks like more or less like an `ld` linker, I'd suggest to avoid adding a new linker flavor for a start and reuse `GccLinker` (especially given that the set of linker flavors accepted by `-C linker-flavor` is a stable public interface).

It is impossible to pass in arguments with spaces to the linker due to the way
how this is implemented for linkers in Rustc. I talked to Alex Crichton
@alexcrichton (hopefully the correct nick) a long time ago and the suggested
way has been to add a new variant. L4-bender is indeed internally using ld
but it `bends` the arguments and adds quite some things and custom options. If
I recall correctly, even the order of the arguments matters, something that
cannot be influenced from outside. One notable one is the `--` that separates
l4-bender arguments from ld arguments that are passed through. How should this
be done?

If a new linker flavour is really not an option, I would ask the compiler team
to resume the discussion on how to quote spaces in linker flags.

bors commented:
:umbrella: The latest upstream changes (presumably #86911) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

pnkfelix commented:
> It is impossible to pass in arguments with spaces to the linker due to the way how this is implemented for linkers in Rustc. I talked to Alex Crichton @alexcrichton (hopefully the correct nick) a long time ago and the suggested way has been to add a new variant. L4-bender is indeed internally using ld but it `bends` the arguments and adds quite some things and custom options. If I recall correctly, even the order of the arguments matters, something that cannot be influenced from outside. One notable one is the `--` that separates l4-bender arguments from ld arguments that are passed through.

I think the points made here seem to give reasonable support for adding a new linker flavor for this case.

humenda commented:
>I think the points made here seem to give reasonable support for adding a new linker flavor for this case.

Could we get some more support, e.g. by a rustc CI,  to detect breakage
earlier? Or is this something that should be done on the L4Re side due to  the
tier 3 support level?

pnkfelix commented:
> I think the points made here seem to give reasonable support for adding a new linker flavor for this case.
> Could we get some more support, e.g. by a rustc CI, to detect breakage earlier? Or is this something that should be done on the L4Re side due to the tier 3 support level?

I think the latter. My understanding of the tier 3 support level is that it is not meant to impose new burdens on the existing infrastructure.

pnkfelix commented:
We [discussed this](https://zulip-archive.rust-lang.org/stream/238009-t-compiler/meetings/topic/.5Bweekly.5D.202021-09-09.20.2354818.html#252636881) in today's T-compiler weekly triage meeting.

Some further review of the [target tier policy](https://doc.rust-lang.org/nightly/rustc/target-tier-policy.html#tier-3-target-policy) led us to [conclude that](https://zulip-archive.rust-lang.org/stream/238009-t-compiler/meetings/topic/.5Bweekly.5D.202021-09-09.20.2354818.html#252648003) an MCP is not strictly required here.

I'm going to double-check the points that @petrochenkov made about the linker, but I'm currently inclined to trust @humenda 's judgement here.

Question for @humenda and @atopia : should either or both of you be recorded as maintainers for the x86_64-unknown-l4re-uclibc target? Is there anyone else who should also be on such a list of maintainers for the target?

petrochenkov commented:
I'll look in more details, this all still look pretty suspicious to me.
I'm not even sure the cmd quoting and env reading stuff belongs to rustc?
There's approximately zero documentation about `l4-bender` on the net (if you have some info about what it does, please share), so it seems niche enough to generally adapt to rustc and not otherwise.

petrochenkov commented:
Here's the version of this PR, but without introducing a new linker flavor - https://github.com/petrochenkov/rust/commit/d3b02f33878c6915d24407c6c747df3168bd244a
It highlights the differences from a regular gcc-style linker, I left questions about them in code.
Most of them are related to the comma vs whitespace story which is still not clear to me.
If `l4-bender` simply passes the arguments after `--` to the underlying linker, then why does e.g. `-z relro` needs to be changed to `-z,relro`?

Also, if there are some issues with passing arguments with whitespaces, then it's better to make a separate issue (with examples from practice, showing how they behave now, and how they should behave instead) and solve it once and for all, instead of introducing more hacks for specific targets.

petrochenkov commented on compiler/rustc_target/src/spec/x86_64_unknown_l4re_uclibc.rs:
```diff
@@ -1,9 +1,13 @@
+    base.crt_static_default = true;
+    base.panic_strategy = PanicStrategy::Abort;
+    base.crt_static_allows_dylibs = false;
+    base.dynamic_linking = false;
```
These options (with exception of `crt_static_default`) specify values that are already used by default or inherited from `l4re_base.rs`, so they can be removed.

petrochenkov commented:
(I'd also suggest submitting the library changes separately, in that case they will get a more appropriate reviewer familiar with library conventions and will likely land faster.)

JohnCSimon commented:
@atopia 
Ping from triage, can you please address the merge conflicts?

atopia commented:
Sorry for the holdup. I'm still unavailable this week but I will address the issues raised in the comments once I'm back to my desk on Monday.

atopia commented:
> Here's the version of this PR, but without introducing a new linker flavor - [petrochenkov@d3b02f3](https://github.com/petrochenkov/rust/commit/d3b02f33878c6915d24407c6c747df3168bd244a) 

I've briefly gave it a run but unsurprisingly it breaks our build system. I am not an expert on `l4-bender`, but without having dug deeply into the matter myself I would rely on @humenda's judgement and the agreement voiced by @pnkfelix. There might be room for improvement in the implementation of the `L4Bender` variant but I would prefer to get Rust running again and refine the support for L4Re later on.

petrochenkov commented:
At this point I'm ok with keeping `l4bender` as a separate linker flavor, if `-Zunstable-options` is required when using `-Clinker-flavor=l4bender` explicitly.

It's clear that we cannot expect the target to be maintained and the linker messiness to be fixed, so let's keep it separate so it can be easily removed when the time comes.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4728006881?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/6109d18d3b22d2f6dd13fe2c96f52ed5ea473e82/checks/4728006881/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking chalk-engine v0.75.0
error: unused import: `PanicStrategy`
 --> compiler/rustc_target/src/spec/x86_64_unknown_l4re_uclibc.rs:1:19
  |
1 | use crate::spec::{PanicStrategy, Target};
  |
  |
  = note: `-D unused-imports` implied by `-D warnings`
    Checking gsgdt v0.1.2
    Checking rls-span v0.5.3
    Checking rls-data v0.19.1
error: could not compile `rustc_target` due to previous error
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4728436580?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/3fa7688908f25eb77c6e15928cfdcf3507697fea/checks/4728436580/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
* highest error code: E0786
Found 502 error codes
Found 0 error codes with no tests
Done!
tidy error: Found 1 features without a gate test.
Expected a gate test for the feature 'l4bender'.
Hint: create a failing test file named 'feature-gate-l4bender.rs'
      in the 'ui' test suite, with its failures due to
      missing usage of `#![feature(l4bender)]`.
Hint: If you already have such a test and don't want to rename it,
      you can also add a // gate-test-l4bender line to the test file.
some tidy checks failed


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "16"


Build completed unsuccessfully in 0:00:13
```

</details>

atopia commented:
> At this point I'm ok with keeping `l4bender` as a separate linker flavor, if `-Zunstable-options` is required when using `-Clinker-flavor=l4bender` explicitly.

Thanks for your feedback. I have split out the library changes to #92612 as requested and addressed most of the FIXME comments in your branch. I still have to see if I can replace reading `L4_BENDER_ARGS` by `-Zpre-link-arg`.

Meanwhile, I have tried to feature gate the linker variant, but it seems I have a major flaw in my understanding: I have tried to follow the instructions given in the Rustc dev guide ([Stability in code](https://rustc-dev-guide.rust-lang.org/implementing_new_features.html#stability-in-code), [Feature gates](https://rustc-dev-guide.rust-lang.org/feature-gates.html)), but I fail to [enable the feature](https://github.com/rust-lang/rust/blob/3fa7688908f25eb77c6e15928cfdcf3507697fea/compiler/rustc_codegen_ssa/src/lib.rs#L9).  I also tried to approach the issue by looking into other active feature gates, but obviously I missed something. Besides, I'm only aware of `cargo`'s  [-Z unstable-options](https://doc.rust-lang.org/cargo/reference/unstable.html) and I don't understand how that would interact with `rustc`.

I'm sorry if I missed something obvious in the documentation, thanks in advance for your patience and any pointers how I could make this work properly.

atopia commented:
@rustbot ready

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4729265851?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/9a56b6735e337a4afbc1a037ebf851a00e0b36a7/checks/4729265851/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
* highest error code: E0786
Found 502 error codes
Found 0 error codes with no tests
Done!
tidy error: /checkout/compiler/rustc_feature/src/active.rs:412: no tracking issue for feature l4bender
tidy error: Found 1 features without a gate test.
Expected a gate test for the feature 'l4bender'.
Hint: create a failing test file named 'feature-gate-l4bender.rs'
      in the 'ui' test suite, with its failures due to
      missing usage of `#![feature(l4bender)]`.
Hint: If you already have such a test and don't want to rename it,
      you can also add a // gate-test-l4bender line to the test file.
some tidy checks failed


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy" "/checkout" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "/checkout/obj/build" "16"


Build completed unsuccessfully in 0:00:12
```

</details>

atopia commented:
> Also, if there are some issues with passing arguments with whitespaces, then it's better to make a separate issue (with examples from practice, showing how they behave now, and how they should behave instead) and solve it once and for all, instead of introducing more hacks for specific targets.

I looked a bit more into your suggestion of using `-Z pre-link-arg`:  the addition of individual `pre_link_arg`s containing whitespace seems fine on the rustc side (my understanding is that [parse_string_push](https://github.com/rust-lang/rust/blob/a77cc64af491a31db224109a76b9b81cd26cd07c/compiler/rustc_session/src/options.rs#L504) is [redirected](https://github.com/rust-lang/rust/blob/a77cc64af491a31db224109a76b9b81cd26cd07c/compiler/rustc_session/src/options.rs#L313) and adds the argument to  `pre_link_args`, from where it is used to build `cmd.args` for the linker invocation without further modification to the individual argument string).

However, the problem is that cargo (understandably) [splits](src/cargo/core/compiler/build_context/target_info.rs) the `CARGO_BUILD_RUSTFLAGS` environment variable on whitespace. Therefore, on our side the choice would have been to lose the benefits of `cargo` or to lose the [build system](https://github.com/kernkonzept/mk) logic around `l4-bender`, or potentially to abuse build.rustflags in cargo's `config.toml`  for dynamic information.

I absolutely agree that parsing a custom environment variable from the linker flavor isn't particularly pretty, but unless there is a major rework of L4Re, I can't think of any other straightforward solution and having an (unstable) linker flavor seems to be the most feasible mid-term way to actually use the l4re target. While L4Re sure is "niche enough" compared to the tier 1 targets, it runs in thousands of EVs and is used in safety-critical and high assurance appliances across various industries.

I'm sorry for neglecting this PR for so long. Right now Rust support for L4Re is still a research project at Kernkonzept but we hope to lay the foundation to attract customer funding for this work. This might entail a larger overhaul of the build system integration that is acceptable for stable rust eventually, but with the current status of the Rust support this is not feasible.

petrochenkov commented:
@atopia 
>or potentially to abuse build.rustflags in cargo's `config.toml`  for dynamic information

That's probably still better than putting this stuff into the compiler for a specific target.

Could you raise this issue on the cargo's Zulip (https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo), maybe people will have some advices.

petrochenkov commented:
>Meanwhile, I have tried to feature gate the linker variant, but it seems I have a major flaw in my understanding

It's much simpler than what the last commit in this PR does, see https://github.com/petrochenkov/rust/commit/18eda458e49d13f2cc08b5672990eadcf224a5ad for an example.

atopia commented:
Thanks for implementing the feature gate @petrochenkov, I really got your request all wrong.

> > or potentially to abuse build.rustflags in cargo's `config.toml`  for dynamic information
> 
> That's probably still better than putting this stuff into the compiler for a specific target.
> 
> Could you raise this issue on the cargo's Zulip (https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo), maybe people will have some advices.

While formulating my request I realized that I had somehow missed `CARGO_ENCODED_RUSTFLAGS` (introduced in  rust-lang/cargo/pull/9601), which solves the build system interfacing issue for `l4-bender`. The last commit removes the custom environment variable parser in L4Bender.

atopia commented:
@rustbot ready

petrochenkov commented:
r=me after squashing commits.

atopia commented:
@rustbot ready

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 29d623528d4c9385c2762fb826b48e4d19470e89 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 29d623528d4c9385c2762fb826b48e4d19470e89 -->
<!-- homu: {"type":"Approved","sha":"29d623528d4c9385c2762fb826b48e4d19470e89","approver":"petrochenkov"} -->
